launch/launch/actions/execute_local.py

```diff
@@ -0,0 +1,663 @@
+
+        cmd = process_event_args['cmd']
+        cwd = process_event_args['cwd']
+        env = process_event_args['env']
+        if self.__log_cmd:
+            self.__logger.info("process details: cmd=[{}], cwd='{}', custom_env?={}".format(
+                ', '.join(cmd), cwd, 'True' if env is not None else 'False'
+            ))
+
+        emulate_tty = self.__emulate_tty
+        if 'emulate_tty' in context.launch_configurations:
+            emulate_tty = evaluate_condition_expression(
+                context,
+                normalize_to_list_of_substitutions(
+                    context.launch_configurations['emulate_tty']
+                ),
+            )
+
+        try:
+            transport, self._subprocess_protocol = await async_execute_process(
+                lambda **kwargs: self.__ProcessProtocol(
+                    self, context, process_event_args, **kwargs
+                ),
+                cmd=cmd,
+                cwd=cwd,
+                env=env,
+                shell=self.__shell,
+                emulate_tty=emulate_tty,
+                stderr_to_stdout=False,
+            )
+        except Exception:
+            self.__logger.error('exception occurred while executing process:\n{}'.format(
+                traceback.format_exc()
+            ))
+            self.__cleanup()
+            return
+
+        pid = transport.get_pid()
+        self._subprocess_transport = transport
+
+        await context.emit_event(ProcessStarted(**process_event_args))
+
+        returncode = await self._subprocess_protocol.complete
+        if returncode == 0:
+            self.__logger.info('process has finished cleanly [pid {}]'.format(pid))
+        else:
+            self.__logger.error("process has died [pid {}, exit code {}, cmd '{}'].".format(
+                pid, returncode, ' '.join(cmd)
+            ))
+        await context.emit_event(ProcessExited(returncode=returncode, **process_event_args))
+        # respawn the process if necessary
+        if not context.is_shutdown and not self.__shutdown_future.done() and self.__respawn:
+            if self.__respawn_delay is not None and self.__respawn_delay > 0.0:
+                # wait for a timeout(`self.__respawn_delay`) to respawn the process
+                # and handle shutdown event with future(`self.__shutdown_future`)
+                # to make sure `ros2 launch` exit in time
+                await asyncio.wait(
+                    [asyncio.sleep(self.__respawn_delay), self.__shutdown_future],
+                    return_when=asyncio.FIRST_COMPLETED
+                )
+            if not self.__shutdown_future.done():
+                context.asyncio_loop.create_task(self.__execute_process(context))
+                return
+        self.__cleanup()
+
+    def prepare(self, context: LaunchContext):
+        """Prepare the action for execution."""
+        self.__process_description.apply_context(context)
+        self.__expand_substitutions(context)
```

Comment:
@roger-strain meta, safe to ignore: I'm not a huge fan of methods with side effects like this. Purity is more expensive but less surprising.